File: engine.py
"""
Matching engine: GitHub profile -> semantic queries -> hybrid issue search -> ranked feed.

Reuses the validated spike logic, but talks to the API directly with a token
(env GITHUB_TOKEN, or `gh auth token` as fallback) so it can run as a server.
"""

import os
import re
import subprocess
import time
from collections import Counter
from datetime import datetime, timezone

import httpx

API = "https://api.github.com"
MIN_STARS = 500
MAX_AGE_DAYS = 365
PER_PAGE = 10
MAX_QUERIES = 5

CONV_PREFIX = re.compile(
    r"^(feat|fix|chore|docs|refactor|perf|test|build|ci|style)(\(.+?\))?!?\s*:\s*", re.I
)
STOPWORDS = set("""
the a an and or of to in for on with from by at is are was were be been this that
it its as into about over after before when while not no yes add adds added adding
update updates updated updating remove removes removed fix fixes fixed use uses used
using make makes made support handle improve implement change changes bump set new
""".split())


# ------------------------------------------------------------------ auth

def get_token():
    tok = os.environ.get("GITHUB_TOKEN")
    if tok:
        return tok
    r = subprocess.run(["gh", "auth", "token"], capture_output=True, text=True)
    if r.returncode == 0 and r.stdout.strip():
        return r.stdout.strip()
    raise RuntimeError(
        "No GitHub token: set GITHUB_TOKEN or run `gh auth login`"
    )


# ------------------------------------------------------------------ api

class GH:
    def __init__(self, token):
        self.c = httpx.Client(
            base_url=API,
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            },
            timeout=30,
        )

    def graphql(self, query):
        r = self.c.post("/graphql", json={"query": query})
        r.raise_for_status()
        data = r.json()
        if data.get("errors"):
            raise RuntimeError(f"GraphQL errors: {data['errors'][:2]}")
        return data

    def search_issues(self, q):
        r = self.c.get(
            "/search/issues",
            params={"q": q, "search_type": "hybrid", "per_page": PER_PAGE},
        )
        r.raise_for_status()
        return r.json()


# ------------------------------------------------------------------ profile

PROFILE_QUERY = """
{
  viewer {
    login
    pullRequests(first: 50, states: MERGED,
                 orderBy: {field: UPDATED_AT, direction: DESC}) {
      nodes {
        title
        repository {
          nameWithOwner
          primaryLanguage { name }
        }
      }
    }
    repositories(first: 30, ownerAffiliations: OWNER, isFork: false,
                 orderBy: {field: PUSHED_AT, direction: DESC}) {
      nodes { primaryLanguage { name } }
    }
  }
}
"""


def build_profile(data):
    viewer = data["data"]["viewer"]
    prs = [p for p in viewer["pullRequests"]["nodes"] if p and p.get("repository")]

    lang_count = Counter()
    for p in prs:
        lang = (p["repository"].get("primaryLanguage") or {}).get("name")
        if lang:
            lang_count[lang] += 1
    for r in viewer["repositories"]["nodes"]:
        lang = (r.get("primaryLanguage") or {}).get("name")
        if lang:
            lang_count[lang] += 2

    keywords = Counter()
    titles = []
    for p in prs:
        t = CONV_PREFIX.sub("", p["title"]).strip()
        if len(t) >= 15:
            titles.append(t)
        for w in re.findall(r"[a-z][a-z0-9\-]{3,}", t.lower()):
            if w not in STOPWORDS:
                keywords[w] += 1

    return {
        "login": viewer["login"],
        "languages": lang_count.most_common(5),
        "keywords": keywords.most_common(10),
        "raw_prs": prs,
    }


def clean_pr_title(title, repo_name):
    t = re.sub(r"#\d+", "", title)
    t = CONV_PREFIX.sub("", t).strip().lstrip(": ").strip()
    repo_tokens = set(re.split(r"[-_/.\s]", repo_name.split("/")[-1].lower()))
    kept = []
    for w in t.split():
        wl = w.lower().strip("():,.'\"")
        if wl in repo_tokens:
            continue
        if len(w) > 1 and (re.search(r"[A-Z]", w[1:]) or w.isupper()):
            continue
        kept.append(w)
    return " ".join(kept).strip()


def build_queries(profile):
    queries = []
    for p in profile["raw_prs"][:6]:
        t = clean_pr_title(p["title"], p["repository"]["nameWithOwner"])
        if len(t.split()) >= 4:
            queries.append({"text": t, "lang": None, "why": f"your PR: '{t[:60]}'"})
        if len(queries) == 3:
            break

    top_langs = [l for l, _ in profile["languages"]]
    top_kw = [k for k, _ in profile["keywords"]]
    if top_langs and len(top_kw) >= 3:
        queries.append({
            "text": f"{top_langs[0]} {' '.join(top_kw[:3])}",
            "lang": top_langs[0],
            "why": f"top language {top_langs[0]} + {top_kw[:3]}",
        })
    if len(top_langs) >= 2 and len(top_kw) >= 6:
        queries.append({
            "text": f"{top_langs[1]} {' '.join(top_kw[3:6])}",
            "lang": top_langs[1],
            "why": f"2nd language {top_langs[1]} + {top_kw[3:6]}",
        })
    return queries[:MAX_QUERIES]


# ------------------------------------------------------------------ scoring

def age_days(iso):
    dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    return (datetime.now(timezone.utc) - dt).days


def score(item, stars):
    s, reasons = 0, []
    labels = {l["name"].lower() for l in item.get("labels", [])}
    days = age_days(item["created_at"])
    comments = item.get("comments", 0)
    repo_fn = item["repository_url"].split("repos/", 1)[-1]
    st = stars.get(repo_fn, 0)

    if any("good first" in l for l in labels):
        s += 3; reasons.append("good-first-issue")
    if any("help wanted" in l for l in labels):
        s += 2; reasons.append("help-wanted")
    if days <= 60:
        s += 2; reasons.append(f"fresh ({days}d)")
    elif days <= 180:
        s += 1; reasons.append(f"recent ({days}d)")
    else:
        s -= 2; reasons.append(f"old ({days}d)")
    if comments <= 10:
        s += 1; reasons.append(f"quiet ({comments} comments)")
    if comments > 30:
        s -= 2; reasons.append(f"heated ({comments} comments)")
    if st >= 1000:
        s += 1; reasons.append(f"popular repo ({st:,}★)")
    return s, reasons


# ------------------------------------------------------------------ feed

def repo_stars(gh, full_names):
    names = full_names[:25]
    fields = []
    for i, fn in enumerate(names):
        owner, name = fn.split("/", 1)
        fields.append(
            f'r{i}: repository(owner: "{owner}", name: "{name}") '
            "{ stargazerCount }"
        )
    if not fields:
        return {}
    data = gh.graphql("{\n" + "\n".join(fields) + "\n}")
    return {
        fn: (data["data"].get(f"r{i}") or {}).get("stargazerCount", 0)
        for i, fn in enumerate(names)
    }


def get_feed(token):
    gh = GH(token)
    profile = build_profile(gh.graphql(PROFILE_QUERY))
    queries = build_queries(profile)

    results = {}
    for q in queries:
        text = f"{q['text']} is:issue is:open no:assignee"
        if q["lang"]:
            text += f" language:{q['lang']}"
        resp = gh.search_issues(text)
        for it in resp.get("items", []):
            user = it.get("user") or {}
            if user.get("type") == "Bot" or user.get("login", "").endswith("[bot]"):
                continue
            results.setdefault(it["id"], (it, set()))[1].add(q["text"][:40])
        time.sleep(0.5)

    repo_names = list({
        it["repository_url"].split("repos/", 1)[-1] for it, _ in results.values()
    })
    stars = repo_stars(gh, repo_names)

    matches = []
    for it, matched in results.values():
        repo_fn = it["repository_url"].split("repos/", 1)[-1]
        if stars.get(repo_fn, 0) < MIN_STARS:
            continue
        if age_days(it["created_at"]) > MAX_AGE_DAYS:
            continue
        s, reasons = score(it, stars)
        if len(matched) >= 2:
            s += 2; reasons.append(f"matched {len(matched)} queries")
        matches.append({
            "score": s,
            "repo": repo_fn,
            "number": it["number"],
            "title": it["title"],
            "url": it["html_url"],
            "age_days": age_days(it["created_at"]),
            "comments": it.get("comments", 0),
            "stars": stars.get(repo_fn, 0),
            "labels": [l["name"] for l in it.get("labels", [])][:4],
            "reasons": reasons,
            "matched": sorted(matched),
        })
    matches.sort(key=lambda m: -m["score"])

    return {
        "profile": profile,
        "queries": queries,
        "matches": matches,
        "built_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }

File: extension/src/engine.ts
/**
 * Matching engine: GitHub profile -> semantic queries -> hybrid issue search -> ranked feed.
 * TypeScript port of engine.py (validated in the spike).
 */
import { graphql, searchIssues } from './github';

const PER_PAGE = 10;
const MAX_QUERIES = 5;

const CONV_PREFIX =
  /^(feat|fix|chore|docs|refactor|perf|test|build|ci|style)(\(.+?\))?!?\s*:\s*/i;
const STOPWORDS = new Set(
  `the a an and or of to in for on with from by at is are was were be been this that
it its as into about over after before when while not no yes add adds added adding
update updates updated updating remove removes removed fix fixes fixed use uses used
using make makes made support handle improve implement change changes bump set new`
    .split(/\s+/)
);

export interface Options {
  minStars: number;
  maxAgeDays: number;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface Pr {
  title: string;
  repository: { nameWithOwner: string; primaryLanguage?: { name: string } | null };
  [k: string]: any;
}

export interface Profile {
  login: string;
  languages: [string, number][];
  keywords: [string, number][];
  prs: Pr[];
}

export interface Query {
  text: string;
  lang: string | null;
  why: string;
}

export interface Match {
  score: number;
  repo: string;
  number: number;
  title: string;
  url: string;
  ageDays: number;
  comments: number;
  stars: number;
  labels: string[];
  reasons: string[];
  matched: string[];
}

export interface Feed {
  profile: Profile;
  queries: Query[];
  matches: Match[];
  builtAt: string;
}

const PROFILE_QUERY = `
{
  viewer {
    login
    pullRequests(first: 50, states: MERGED,
                 orderBy: {field: UPDATED_AT, direction: DESC}) {
      nodes {
        title
        repository {
          nameWithOwner
          primaryLanguage { name }
        }
      }
    }
    repositories(first: 30, ownerAffiliations: OWNER, isFork: false,
                 orderBy: {field: PUSHED_AT, direction: DESC}) {
      nodes { primaryLanguage { name } }
    }
  }
}
`;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function buildProfile(data: any): Profile {
  const viewer = data.data.viewer;
  const prs: Pr[] = (viewer.pullRequests.nodes || []).filter(
    (p: Pr | null) => p && p.repository
  );

  const langCount = new Map<string, number>();
  for (const p of prs) {
    const lang = p.repository.primaryLanguage?.name;
    if (lang) langCount.set(lang, (langCount.get(lang) ?? 0) + 1);
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  for (const r of viewer.repositories.nodes || []) {
    const lang = r.primaryLanguage?.name;
    if (lang) langCount.set(lang, (langCount.get(lang) ?? 0) + 2);
  }

  const kw = new Map<string, number>();
  for (const p of prs) {
    const t = p.title.replace(CONV_PREFIX, '').trim();
    for (const w of t.toLowerCase().match(/[a-z][a-z0-9-]{3,}/g) ?? []) {
      if (!STOPWORDS.has(w)) kw.set(w, (kw.get(w) ?? 0) + 1);
    }
  }

  const byCount = (a: [string, number], b: [string, number]) => b[1] - a[1];
  return {
    login: viewer.login,
    languages: [...langCount.entries()].sort(byCount).slice(0, 5),
    keywords: [...kw.entries()].sort(byCount).slice(0, 10),
    prs,
  };
}

function cleanPrTitle(title: string, repoName: string): string {
  let t = title.replace(/#\d+/g, '');
  t = t.replace(CONV_PREFIX, '').trim().replace(/^:\s*/, '').trim();
  const repoTokens = new Set(
    repoName.split('/')[1]?.toLowerCase().split(/[-_/.\s]/) ?? []
  );
  const kept: string[] = [];
  for (const w of t.split(/\s+/)) {
    const wl = w.toLowerCase().replace(/[():,.'"]/g, '');
    if (repoTokens.has(wl)) continue;
    if (w.length > 1 && (/[A-Z]/.test(w.slice(1)) || w === w.toUpperCase())) continue;
    kept.push(w);
  }
  return kept.join(' ').trim();
}

export function buildQueries(profile: Profile): Query[] {
  const queries: Query[] = [];
  for (const p of profile.prs.slice(0, 6)) {
    const t = cleanPrTitle(p.title, p.repository.nameWithOwner);
    if (t.split(/\s+/).filter(Boolean).length >= 4) {
      queries.push({ text: t, lang: null, why: `your PR: '${t.slice(0, 60)}'` });
    }
    if (queries.length === 3) break;
  }

  const topLangs = profile.languages.map(([l]) => l);
  const topKw = profile.keywords.map(([k]) => k);
  if (topLangs.length && topKw.length >= 3) {
    queries.push({
      text: `${topLangs[0]} ${topKw.slice(0, 3).join(' ')}`,
      lang: topLangs[0],
      why: `top language ${topLangs[0]} + ${topKw.slice(0, 3).join(', ')}`,
    });
  }
  if (topLangs.length >= 2 && topKw.length >= 6) {
    queries.push({
      text: `${topLangs[1]} ${topKw.slice(3, 6).join(' ')}`,
      lang: topLangs[1],
      why: `2nd language ${topLangs[1]} + ${topKw.slice(3, 6).join(', ')}`,
    });
  }
  return queries.slice(0, MAX_QUERIES);
}

function ageDays(iso: string): number {
  return Math.floor((Date.now() - new Date(iso).getTime()) / 86_400_000);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function score(item: any, stars: Map<string, number>): [number, string[]] {
  let s = 0;
  const reasons: string[] = [];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const labels = new Set<string>((item.labels ?? []).map((l: any) => l.name.toLowerCase()));
  const days = ageDays(item.created_at);
  const comments: number = item.comments ?? 0;
  const repoFn = item.repository_url.split('repos/')[1];
  const st = stars.get(repoFn) ?? 0;

  if ([...labels].some((l) => l.includes('good first'))) {
    s += 3; reasons.push('good-first-issue');
  }
  if ([...labels].some((l) => l.includes('help wanted'))) {
    s += 2; reasons.push('help-wanted');
  }
  if (days <= 60) { s += 2; reasons.push(`fresh (${days}d)`); }
  else if (days <= 180) { s += 1; reasons.push(`recent (${days}d)`); }
  else { s -= 2; reasons.push(`old (${days}d)`); }
  if (comments <= 10) { s += 1; reasons.push(`quiet (${comments} comments)`); }
  if (comments > 30) { s -= 2; reasons.push(`heated (${comments} comments)`); }
  if (st >= 1000) { s += 1; reasons.push(`popular repo (${st.toLocaleString()}★)`); }
  return [s, reasons];
}

async function repoStars(
  token: string,
  fullNames: string[]
): Promise<Map<string, number>> {
  const names = fullNames.slice(0, 25);
  if (!names.length) return new Map();
  const fields = names.map((fn, i) => {
    const [owner, name] = fn.split('/');
    return `r${i}: repository(owner: "${owner}", name: "${name}") { stargazerCount }`;
  });
  const data = await graphql(token, `{\n${fields.join('\n')}\n}`);
  const out = new Map<string, number>();
  names.forEach((fn, i) => {
    out.set(fn, data.data[`r${i}`]?.stargazerCount ?? 0);
  });
  return out;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export async function getFeed(token: string, opts: Options): Promise<Feed> {
  const profile = buildProfile(await graphql(token, PROFILE_QUERY));
  const queries = buildQueries(profile);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const results = new Map<number, { item: any; matched: Set<string> }>();
  for (const q of queries) {
    let text = `${q.text} is:issue is:open no:assignee`;
    if (q.lang) text += ` language:${q.lang}`;
    const resp = await searchIssues(token, text, PER_PAGE);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    for (const it of resp.items ?? []) {
      const login: string = it.user?.login ?? '';
      if (it.user?.type === 'Bot' || login.endsWith('[bot]')) continue;
      const entry = results.get(it.id) ?? { item: it, matched: new Set<string>() };
      entry.matched.add(q.text.slice(0, 40));
      results.set(it.id, entry);
    }
    await sleep(500); // semantic search: 10 req/min
  }

  const repoNames = [
    ...new Set(
      [...results.values()].map((r) => r.item.repository_url.split('repos/')[1] as string)
    ),
  ];
  const stars = await repoStars(token, repoNames);

  const matches: Match[] = [];
  for (const { item, matched } of results.values()) {
    const repoFn = item.repository_url.split('repos/')[1] as string;
    const st = stars.get(repoFn) ?? 0;
    if (st < opts.minStars) continue;
    const days = ageDays(item.created_at);
    if (days > opts.maxAgeDays) continue;
    let [s, reasons] = score(item, stars);
    if (matched.size >= 2) {
      s += 2;
      reasons = [...reasons, `matched ${matched.size} queries`];
    }
    matches.push({
      score: s,
      repo: repoFn,
      number: item.number,
      title: item.title,
      url: item.html_url,
      ageDays: days,
      comments: item.comments ?? 0,
      stars: st,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      labels: (item.labels ?? []).map((l: any) => l.name as string).slice(0, 4),
      reasons,
      matched: [...matched].sort(),
    });
  }
  matches.sort((a, b) => b.score - a.score);

  return {
    profile,
    queries,
    matches,
    builtAt: new Date().toISOString(),
  };
}
